File: go/proverb/proverb.go
package proverb

import (
	"fmt"
)

var PROVERB_LINE_FORMAT = "For want of a %s the %s was lost."
var FINAL_LINE_FORMAT = "And all for the want of a %s."

func Proverb(rhyme []string) []string {
	if len(rhyme) == 0 {
		return []string{}
	}

	if len(rhyme) == 1 {
		return []string{fmt.Sprintf(FINAL_LINE_FORMAT, rhyme[0])}
	}

	proverb := make([]string, 0)

	for i := 0; i < len(rhyme) - 1; i++ {
		proverb = append(proverb, fmt.Sprintf(PROVERB_LINE_FORMAT, rhyme[i], rhyme[i + 1]))
	}

	proverb = append(proverb, fmt.Sprintf(FINAL_LINE_FORMAT, rhyme[0]))

	return proverb
}

File: java/proverb/src/main/java/Proverb.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Proverb
{
    private final String[] words;
    private static final String PROVERB_LINE_FORMAT = "For want of a %s the %s was lost.\n";
    private static final String FINAL_LINE_FORMAT = "And all for the want of a %s.";

    Proverb(String[] words)
    {
        this.words = words;
    }

    String recite()
    {
        StringBuilder proverbBuilder = new StringBuilder();
        String proverb = IntStream
                .range(0, words.length - 1)
                .mapToObj(i -> String.format(PROVERB_LINE_FORMAT, words[i], words[i + 1]))
                .collect(Collectors.joining());

        return words.length == 0
                ? ""
                : proverbBuilder.append(proverb).append(String.format(FINAL_LINE_FORMAT, words[0])).toString();
    }
}
